source/MaterialXGenShader/UnitSystem.cpp

```diff
@@ -14,30 +14,43 @@
 namespace MaterialX
 {
 
-// Helper class to create the constant block for mx_<unittype>_unit
-class ScalarUnitNode : public SourceCodeNode
+class ScalarUnitNode : public ShaderNodeImpl
 {
   public:
     explicit ScalarUnitNode(LinearUnitConverterPtr scalarUnitConverter) :
-        _scalarUnitConverter(scalarUnitConverter)
+        _scalarUnitConverter(scalarUnitConverter),
+        _unitRatioFunctionName("mx_" + _scalarUnitConverter->getUnitType() + "_unit_ratio")
     {
     }
 
     static ShaderNodeImplPtr create(LinearUnitConverterPtr scalarUnitConverter);
 
+    void initialize(const InterfaceElement& element, GenContext& context) override;
     void emitFunctionDefinition(const ShaderNode& node, GenContext& context, ShaderStage& stage) const override;
+    void emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderStage& stage) const override;
 
   protected:
     LinearUnitConverterPtr _scalarUnitConverter;
+    const string _unitRatioFunctionName;
 };
 
 ShaderNodeImplPtr ScalarUnitNode::create(LinearUnitConverterPtr scalarUnitConverter)
 {
     return std::make_shared<ScalarUnitNode>(scalarUnitConverter);
 }
 
-void ScalarUnitNode::emitFunctionDefinition(const ShaderNode& node, GenContext& context, ShaderStage& stage) const
+void ScalarUnitNode::initialize(const InterfaceElement& element, GenContext& /*context*/)
 {
+    _name = element.getName();
+
+    // Use the unit ratio function name has hash to make sure this function
+    // is shared, and only emitted once, for all units of the same unit type.
+    _hash = std::hash<string>{}(_unitRatioFunctionName);
```

Comment:
Forgot about the hash for this. 

@niklasharrysson for an example for testing multiple angle units, this can be added to `TestSuite/stdlib/math/vector_math.mtlx`
```
 <nodegraph name="rotate_angles" type="" xpos="7.24435" ypos="23.808">
    <position name="position1" type="vector3" xpos="3.53793" ypos="5.42">
      <parameter name="space" type="string" value="object" />
    </position>
    <rotate name="rotate1" type="vector3" xpos="5.19248" ypos="5.306">
      <input name="in" type="vector3" value="0.0, 0.0, 0.0" nodename="position1" />
      <input name="amount" type="float" value="180.0000" unittype="angle" unit="degrees" />
      <parameter name="axis" type="vector3" value="0.0, 1.0000, 1.0000" />
    </rotate>
    <rotate name="rotate2" type="vector3" xpos="5.19248" ypos="5.306">
      <input name="in" type="vector3" value="0.0, 0.0, 0.0" nodename="position1" />
      <input name="amount" type="float" value="1.5708" unittype="angle" unit="radians" />
      <parameter name="axis" type="vector3" value="0.0, 1.0000, 1.0000" />
    </rotate>
    <multiply name="multiply1" type="vector3" xpos="5.74483" ypos="4.58">
      <input name="in1" type="vector3" value="1.0000, 0.5000, 0.0" nodename="rotate1"/>
      <input name="in2" type="vector3" value="0.5000, 0.5000, 1.0000" nodename="rotate2"/>
    </multiply>
    <output name="out" type="vector3" nodename="multiply1" />
  </nodegraph>
```